### PR TITLE
document modbus fp32 datatype

### LINKED ISSUE #266: Compile error on 1.13.0-dev in src/esphome/core/automation.h (works on 1.12.2)
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Sonoff S31 Hass.io v1.13.0-dev (not working) WSL on Win10 v1.12.2 (working)
**ESP (ESP32/ESP8266, Board/Sonoff):**

**Affected component:**

Generic Sonoff / S31 with Voltage Monitoring.

**Description of problem:**
Errors trying to compile from within Hass.io Dashboard.  See error below.  Same yaml code compiles and works fine with 1.12.2 when compiled from a Windows machine.

**Problem-relevant YAML-configuration entries:**
```yaml
PASTE YAML FILE HERE
esphome:
  name: washing_machine
  platform: ESP8266
  board: esp01_1m

wifi:
  ssid: "<REDACTED>"
  password: "<REDACTED>"

api:

logger:
  baud_rate: 0

ota:

binary_sensor:
  - platform: gpio
    pin:
      number: GPIO0
      mode: INPUT_PULLUP
      inverted: True
    name: "Washer Button"
    on_press:
      - switch.toggle: relay
  - platform: status
    name: "Washer Status"

switch:
  - platform: gpio
    name: "Washer Relay"
    pin: GPIO12
    id: relay

output:
  # Register the green LED as a dimmable output ....
  - platform: esp8266_pwm
    id: s31_blue_led
    pin:
      number: GPIO13
      inverted: True

light:
  # ... and then make a light out of it.
  - platform: monochromatic
    name: "Washer Blue LED"
    output: s31_blue_led

uart:
  rx_pin: RX
  baud_rate: 4800

sensor:
  - platform: cse7766
    current:
      name: "Washer Current"
    voltage:
      name: "Washer Voltage"
    power:
      name: "Washer Power"
```

**Traceback (if applicable):**

```
Compiling /data/washing_machine/.pioenvs/washing_machine/lib4d9/ESP8266WiFi/ESP8266WiFiGeneric.cpp.o
In file included from src/esphome/components/api/service_call_message.h:4:0,
from src/esphome/components/api/api_server.h:14,
from src/main.cpp:4:
src/esphome/core/automation.h: In instantiation of 'bool esphome::Trigger<Ts>::is_running() [with Ts = {}]':
src/esphome/components/light/base_light_effects.h:72:34:   required from here
src/esphome/core/automation.h:58:48: error: request for member 'is_running' in '((esphome::Trigger<>*)this)->esphome::Trigger<>::automation_parent_', which is of pointer type 'esphome::Automation<>*' (maybe you meant to use '->' ?)
return this->automation_parent_.is_running();
^
Compiling /data/washing_machine/.pioenvs/washing_machine/lib4d9/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Compiling /data/washing_machine/.pioenvs/washing_machine/lib4d9/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
src/esphome/core/automation.h: In member function 'bool esphome::Trigger<Ts>::is_running() [with Ts = {}]':
src/esphome/core/automation.h:59:3: warning: control reaches end of non-void function [-Wreturn-type]
}
^
*** [/data/washing_machine/.pioenvs/washing_machine/src/main.cpp.o] Error 1
```

**Additional information and things you've tried:**

### PR DESCRIPTION
**Pull request by martgras**
## Description:

Add missing documentation about FP32 data type 
Improve serial logging information
Add usage details for register_count
Add documentation for custom_command

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#2680

## Checklist:

  - [x] Branch: `next` is for changes and new documentation that will go public with the next ESPHome release. Fixes, changes and adjustments for the current release should be created against `current`.
  - [ ] Link added in `/index.rst` when creating new documents for new components or cookbook.

### CODE REVIEW

File: components/modbus_controller.rst [hunk 1]
```diff
@@ -53,73 +68,73 @@ Technically there is no difference between the "inline" and the standard definit
 .. code-block:: yaml
 
     esphome:
-      name: solarstation
-      platform: ESP32
-      board: esp32dev
+        name: solarstation
+        platform: ESP32
+        board: esp32dev
```
Comment (jesserockz) [inline components/modbus_controller.rst]: No need to ad extra spaces here? (Should be 2 spaces for yaml)

**Comment by jesserockz** [comment]
As this is a feature matched with a PR in `esphome/esphome`, please target your PR to the `next` branch and rebase.

Thanks
Jesse

File: components/modbus_controller.rst (hunk 1 above)
Comment (martgras) [reply components/modbus_controller.rst]: right - the linter errors about the 4 spaces  confused me.
Branch is changed to next now

File: components/modbus_controller.rst [hunk 2]
```diff
@@ -29,6 +25,25 @@ See [How is this RS485 Module Working?](https://electronics.stackexchange.com/qu
 
 The controller connects to the UART of the MCU. For ESP32  GPIO PIN 16 to TXD PIN 17 to RXD are the default ports but any other pins can be used as well . 3.3V to VCC and GND to GND.
 
+.. note::
+
+    If your are using an ESP82xx serial logging may cause problems reading from uart. For best results hardware serial is prefered. Software serial may not be able to read all received data if other components spend a lot of time in loop()
```
Comment (jesserockz) [inline components/modbus_controller.rst]:
```suggestion
    If you are using an ESP8266, serial logging may cause problems reading from UART. For best results, hardware serial is recommended. Software serial may not be able to read all received data if other components spend a lot of time in the ``loop()``.
```

File: components/modbus_controller.rst [hunk 3]
```diff
@@ -29,6 +25,25 @@ See [How is this RS485 Module Working?](https://electronics.stackexchange.com/qu
...
+    
+    For hardware serial only a limited set of pins can be used. Either `tx_pin: GPIO1` and `rx_pin: GPIO3`  or `tx_pin: GPIO15` and `rx_pin: GPIO13`.
+    
+    The disadvantage of using the hardware uart is that you can't use serial logging because the serial logs would be sent to the modbus device and cause errors.
+    
+    Serial logging can be disable by setting `baud_rate: 0`.
```
Comment (jesserockz) [inline components/modbus_controller.rst]:
```suggestion
    Serial logging can be disabled by setting `baud_rate: 0`.
```

File: components/switch/modbus_controller.rst [hunk 4]
```diff
@@ -22,6 +22,49 @@ Configuration variables:
   To get the value of the coil register 2 can be retrived using address: 2 / offset: 0 or address: 0 / offset 2
 - **bitmask** : some values are packed in a response. The bitmask is used to determined if the result is true or false
 - **skip_updates**: (*Optional*, integer): By default all sensors of of a modbus_controller are updated together. For data points that don't change very frequently updates can be skipped. A value of 5 would only update this sensor range in every 5th update cycle
+- **use_write_multiple**: (*Optional*, boolean): By default the modbus command `Force Single Coil` (function code 5) is used to send state changes to the device. If your device onyl supports `Force Multiple Coils` (function code 15) set this option to true.
```
Comment (jesserockz) [inline components/switch/modbus_controller.rst]:
```suggestion
- **use_write_multiple**: (*Optional*, boolean): By default the modbus command `Force Single Coil` (function code 5) is used to send state changes to the device. If your device only supports `Force Multiple Coils` (function code 15) set this option to true.
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
All things like: 

```rst
If `custom_data` is used
```

Should be:

```rst
If ``custom_data`` is used
```

Double backticks to wrap variables etc

**Comment by martgras** [comment]
I hope I replace all instances where single backtick was used with the last commit

File: components/sensor/modbus_controller.rst [hunk 5]
```diff
@@ -14,29 +14,38 @@ Configuration variables:
 - **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
 - **name** (**Required**, string): The name of the sensor.
 - **register_type** (**Required**): type of the modbus register.
-    - coil: coils are also called discrete outout. Coils are 1-bit registers (on/off values) that are used to control discrete outputs. Read and Write access
-    - discrete_input: discrete input register (read only coil) are similar to coils but can only be read.
-    - holding: Holding Registers - Holding registers are the most universal 16-bit register. Read and Write access
-    - read: Read Input Registers - registers are 16-bit registers used for input, and may only be read
+    - coil: coils are also called discrete outout. Coils are 1-bit registers (on/off values) that are used to control discrete outputs. Read and Write access. Modbus function code 1 (Read Coil Status) will be used
```
Comment (jesserockz) [inline components/sensor/modbus_controller.rst]:
```suggestion
    - coil: coils are also called discrete output. Coils are 1-bit registers (on/off values) that are used to control discrete outputs. Read and Write access. Modbus function code 1 (Read Coil Status) will be used
```

File: components/sensor/modbus_controller.rst [hunk 6]
```diff
@@ -119,14 +128,138 @@ Parameters passed into the lambda
...
 
  - ``return <FLOATING_POINT_NUMBER>;`` the new value for the sensor.
  - ``return NAN;`` if the state should be considered invalid to indicate an error (advanced).
 
+.. _modbus_custom_data:
+
+Using custom_data
+-----------------
+
+`custom_data` can be used to create an arbitrary modbus command. Combined with a lambda any response can be handled.
```
Comment (jesserockz) [inline components/sensor/modbus_controller.rst]:
```suggestion
``custom_data`` can be used to create an arbitrary modbus command. Combined with a lambda any response can be handled. 
```

File: components/sensor/modbus_controller.rst [hunk 7]
```diff
@@ -119,14 +128,138 @@ Parameters passed into the lambda
...
+              raw_to_float.raw =   data[0] << 24 | data[1] << 16 | data[2] << 8 |  data[3];
+              ESP_LOGD("Modbus Sensor Lambda", "FP32 = 0x%08X => %f", raw_to_float.raw, raw_to_float.float_value);
+              return raw_to_float.float_value;
+            unit_of_measurement: kVArh
+            accuracy_decimals: 1
+
+
+.. note:: **Optimize modbus communications**
+
+    `register_count` can also be used to skip a register in consecutive range.
```
Comment (jesserockz) [inline components/sensor/modbus_controller.rst]:
```suggestion
    ``register_count`` can also be used to skip a register in consecutive range. 
```

File: components/binary_sensor/modbus_controller.rst [hunk 8]
```diff
@@ -23,7 +23,10 @@ Configuration variables:
 - **address**: (**Required**, integer): start address of the first register in a range
 - **bitmask** : some values are packed in a response. The bitmask is used to determined if the result is true or false
 - **skip_updates**: (*Optional*, integer): By default all sensors of of a modbus_controller are updated together. For data points that don't change very frequently updates can be skipped. A value of 5 would only update this sensor range in every 5th update cycle
-- **force_new_range**: (*Optional*, boolean): If possible sensors with sequential addresses are grouped together and requested in one range. Setting `foce_new_range: true` enforces the start of a new range at that address.
+- **force_new_range**: (*Optional*, boolean): If possible sensors with sequential addresses are grouped together and requested in one range. Setting ``force_new_range: true`` enforces the start of a new range at that address.
+- **custom_data** (**Optional**, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used.
```
Comment (jesserockz) [inline components/binary_sensor/modbus_controller.rst]:
```suggestion
- **custom_data** (*Optional*, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used. 
```

File: components/number/modbus_controller.rst [hunk 9]
```diff
@@ -19,27 +19,32 @@ Configuration variables:
...
+  The number of registers this data point spans. Overrides the defaults determined by ``value_type``.
+- **force_new_range**: (*Optional*, boolean): If possible sensors with sequential addresses are grouped together and requested in one range. Setting ``force_new_range: true`` enforces the start of a new range at that address.
 - **offset**: (*Optional*, int): only required for uncommon response encodings
     offset from start address in bytes. If more than one register is read a modbus read registers command this value is used to find the start of this datapoint relative to start address. The component calculates the size of the range based on offset and size of the value type
 - **min_value** (*Optional*, float): The minimum value this number can be.
 - **max_value** (*Optional*, float): The maximum value this number can be.
 - **step** (*Optional*, float): The granularity with which the number can be set. Defaults to 1
+- **custom_data** (**Optional**, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used.
```
Comment (jesserockz) [inline components/number/modbus_controller.rst]:
```suggestion
- **custom_data** (*Optional*, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used.
```

File: components/sensor/modbus_controller.rst [hunk 10]
```diff
@@ -14,29 +14,38 @@ Configuration variables:
...
-- **force_new_range**: (*Optional*, boolean): If possible sensors with sequential addresses are grouped together and requested in one range. Setting `foce_new_range: true` enforces the start of a new range at that address.
+  The number of registers this data point spans. Overrides the defaults determined by ``value_type``.
+  If no value for ``register_count`` is provided, it is calculated based on the register type.
+
+  The default size for 1 register is 16 bits (1 Word). Some devices are not adhering to this convention and have registers larger than 16 bits.  In this case ``register_count`` must be set. For example, if your modbus device uses 1 registers for a FP32 value instead the default of two set ``register_count: 2``.
+
+- **force_new_range**: (*Optional*, boolean): If possible sensors with sequential addresses are grouped together and requested in one range. Setting ``force_new_range: true`` enforces the start of a new range at that address.
+- **custom_data** (**Optional**, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used.
```
Comment (jesserockz) [inline components/sensor/modbus_controller.rst]:
```suggestion
- **custom_data** (*Optional*, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used. 
```

File: components/switch/modbus_controller.rst [hunk 11]
```diff
@@ -22,6 +22,49 @@ Configuration variables:
   To get the value of the coil register 2 can be retrived using address: 2 / offset: 0 or address: 0 / offset 2
 - **bitmask** : some values are packed in a response. The bitmask is used to determined if the result is true or false
 - **skip_updates**: (*Optional*, integer): By default all sensors of of a modbus_controller are updated together. For data points that don't change very frequently updates can be skipped. A value of 5 would only update this sensor range in every 5th update cycle
+- **use_write_multiple**: (*Optional*, boolean): By default the modbus command ``Force Single Coil`` (function code 5) is used to send state changes to the device. If your device only supports ``Force Multiple Coils`` (function code 15) set this option to true.
+- **custom_data** (**Optional**, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used.
```
Comment (jesserockz) [inline components/switch/modbus_controller.rst]:
```suggestion
- **custom_data** (*Optional*, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used. 
```

File: components/text_sensor/modbus_controller.rst [hunk 12]
```diff
@@ -26,9 +26,12 @@ Configuration variables:
 - **raw_encode**: (*Optional*, NONE , HEXBYTES, COMMA) If the response is binary it can't be published directly. Since a text sensor only publishes strings the binary data can encoded
      - HEXBYTES:  2 byte hex string. 0x2011 will be sent as "2011".
      - COMMA: Byte values as integers, delimited by a coma. 0x2011 will be sent as "32,17"
-- **force_new_range**: (*Optional*, boolean): If possible sensors with sequential addresses are grouped together and requested in one range. Setting `foce_new_range: true` enforces the start of a new range at that address.
+- **force_new_range**: (*Optional*, boolean): If possible sensors with sequential addresses are grouped together and requested in one range. Setting ``force_new_range: true`` enforces the start of a new range at that address.
+- **custom_data** (**Optional**, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used.
```
Comment (jesserockz) [inline components/text_sensor/modbus_controller.rst]:
```suggestion
- **custom_data** (*Optional*, list of bytes): raw bytes for modbus command. This allows using non-standard commands. If ``custom_data`` is used ``address`` and ``register_type`` can't be used. 
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**